core/unix/coredump.c

```diff
@@ -186,16 +182,17 @@ os_dump_core_internal(void)
     // the section name in the section name table.
     while (memquery_iterator_next(&iter)) {
         // Skip non-readable section.
-        if (iter.prot == MEMPROT_NONE || strcmp(iter.comment, VVAR_SECTION) == 0) {
+        if (iter.prot == MEMPROT_NONE || strcmp(iter.comment, VVAR_SECTION) == 0 ||
+            strcmp(iter.comment, VSYSCALL_SECTION) == 0) {
             continue;
         }
         ELF_ADDR offset = 0;
         if (iter.comment != NULL && iter.comment[0] != '\0') {
-            offset = (ELF_ADDR)strhash_hash_lookup(GLOBAL_DCONTEXT, string_htable,
-                                                   iter.comment);
+            const char *pos = strstr(&string_table[1], iter.comment);
```

Comment:
How large does this table get?  If not very large then a linear scan is probably ok.  If it gets large: go back to the hashtable with no lock b/c we're single-threaded here, having suspended all others?  Actually you didn't pass HASHTABLE_SHARED for your flags: so shouldn't there have been no lock on the table anyway?

Reply:
The table is about 4k bytes with a simple application.

I added HASHTABLE_SHARED after running into the lock issue with debug build, and it still hit the lock order issue. Hence I changed it to strstr().

Reply:
Let me check if it makes a difference if the function is not called from an event callback.

Reply:
No, HASHTABLE_SHARED is what adds a lock: see TABLE_NEEDS_LOCK in hashtable.h.

Reply:
I see. The original version has table_flags = 0:

   strhash_table_t *string_htable =
        strhash_hash_create(GLOBAL_DCONTEXT, /*bits=*/8, /*load_factor_percent=*/80,
                            /*table_flags=*/0, NULL _IF_DEBUG("mmap-string-table"));

and I saw the rack order violation error with the debug build. 

I've tried the hashtable with event_nudge, and I got the same rang order violation error:

<rank order violation global_alloc_lock(recursive)@/usr/local/google/home/kyluk/dynamorio-x86/dynamorio/core/heap.c:314 acquired after memory_info_buf_lock(mutex)@/usr/local/google/home/kyluk/dynamorio-x86/dynamorio/core/unix/memquery_linux.c:70 in tid:3a93f7>
<rank order violation report_buf_lock(mutex)@/usr/local/google/home/kyluk/dynamorio-x86/dynamorio/core/utils.c:2015 acquired after memory_info_buf_lock(mutex)@/usr/local/google/home/kyluk/dynamorio-x86/dynamorio/core/unix/memquery_linux.c:70 in tid:3a93f7>
<Application /usr/local/google/home/kyluk/dynamorio-x86/dynamorio/build/suite/tests/bin/simple_app (3838967).  Internal Error: DynamoRIO debug check failure: /usr/local/google/home/kyluk/dynamorio-x86/dynamorio/core/utils.c:615 (dcontext->thread_owned_locks->last_lock->rank < lock->rank || first_client || both_client) && "rank order violation"
(Error occurred @1 frags in tid 3838967)

Reply:
I've put back the hash table, and use it with a lock.

Also added an ASSERT to ensure only 1 thread is running.